File: frontend/lib/services/socket_service.dart
import 'dart:async';
// ignore: library_prefixes
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../core/constants/api_config.dart';

class DangerZoneAlert {
  final String zoneId;
  final String zoneName;
  final int dangerLevel;
  final String eventType;
  final String message;

  DangerZoneAlert.fromJson(Map<String, dynamic> json)
      : zoneId = json['zoneId']?.toString() ?? '',
        zoneName = json['zoneName']?.toString() ?? 'Vùng nguy hiểm',
        dangerLevel = (json['dangerLevel'] as num?)?.toInt() ?? 1,
        eventType = json['eventType']?.toString() ?? '',
        message = json['message']?.toString() ?? 'Cảnh báo vùng nguy hiểm!';
}

class SocketService {
  IO.Socket? _socket;
  final _emergencyController = StreamController<bool>.broadcast();
  final _dangerZoneAlertController =
      StreamController<DangerZoneAlert>.broadcast();

  Stream<bool> get onEmergencyChange => _emergencyController.stream;
  Stream<DangerZoneAlert> get onDangerZoneAlert =>
      _dangerZoneAlertController.stream;

  void connect() {
    _socket = IO.io(ApiConfig.socketUrl, <String, dynamic>{
      'transports': ['websocket'],
      'autoConnect': false,
      'reconnection': true,
      'reconnectionDelay': 5000,
      'reconnectionAttempts': 3,
    })
      ..on('air_raid_alert', (_) => _emergencyController.add(true))
      ..on('safe_alert', (_) => _emergencyController.add(false))
      ..on('entered_danger_zone', (_) => _emergencyController.add(true))
      ..on('exited_danger_zone', (_) => _emergencyController.add(false))
      ..on('danger_zone_alert', (data) {
        try {
          _dangerZoneAlertController.add(
            DangerZoneAlert.fromJson(Map<String, dynamic>.from(data as Map)),
          );
        } catch (_) {}
      })
      ..connect();
  }

  void updateLocation(double lat, double lng) {
    _socket?.emit('register_location', {'lat': lat, 'lng': lng});
  }

  void dispose() {
    _socket?.dispose();
    _emergencyController.close();
    _dangerZoneAlertController.close();
  }
}

File: frontend-rn/src/services/socket.service.ts
import { io, type Socket } from 'socket.io-client';
import { API_CONFIG } from '@/config/api';

export interface DangerZoneAlert {
  zoneId: string;
  zoneName: string;
  dangerLevel: number;
  eventType: string;
  message: string;
}

export interface SocketHandlers {
  onEmergencyChange: (isEmergency: boolean) => void;
  onDangerZoneAlert: (alert: DangerZoneAlert) => void;
}

export class SocketService {
  private socket: Socket | null = null;
  private expoPushToken: string | null = null;

  connect(h: SocketHandlers) {
    this.socket = io(API_CONFIG.socketUrl, {
      transports: ['websocket'],
      autoConnect: false,
      reconnection: true,
      reconnectionDelay: 5000,
      reconnectionAttempts: 3,
    });
    this.socket.on('air_raid_alert', () => h.onEmergencyChange(true));
    this.socket.on('safe_alert', () => h.onEmergencyChange(false));
    this.socket.on('entered_danger_zone', () => h.onEmergencyChange(true));
    this.socket.on('exited_danger_zone', () => h.onEmergencyChange(false));
    this.socket.on('danger_zone_alert', (data: any) => {
      try {
        h.onDangerZoneAlert({
          zoneId: String(data?.zoneId ?? ''),
          zoneName: String(data?.zoneName ?? 'Vùng nguy hiểm'),
          dangerLevel: Number(data?.dangerLevel ?? 1),
          eventType: String(data?.eventType ?? ''),
          message: String(data?.message ?? 'Cảnh báo vùng nguy hiểm!'),
        });
      } catch {
        /* ignore */
      }
    });
    this.socket.connect();
  }

  setExpoPushToken(token: string | null) {
    this.expoPushToken = token;
  }

  updateLocation(lat: number, lng: number) {
    this.socket?.emit('register_location', {
      lat,
      lng,
      expoPushToken: this.expoPushToken ?? undefined,
    });
  }

  dispose() {
    this.socket?.disconnect();
    this.socket = null;
  }
}
